File: pkg/persist/persist_test.go
package persist

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"
)

type testDoc struct {
	Name string `json:"name"`
	N    int    `json:"n"`
}

func TestSaveLoadJSON_RoundTrip(t *testing.T) {
	path := filepath.Join(t.TempDir(), "doc.json")

	if err := Save(path, &testDoc{Name: "Ash Prime", N: 3}); err != nil {
		t.Fatalf("Save: %v", err)
	}

	var got testDoc
	if err := LoadJSON(path, &got); err != nil {
		t.Fatalf("LoadJSON: %v", err)
	}
	if got.Name != "Ash Prime" || got.N != 3 {
		t.Errorf("LoadJSON = %+v, want {Name:Ash Prime N:3}", got)
	}
}

func TestLoadJSON_MissingFilePreservesIsNotExist(t *testing.T) {
	path := filepath.Join(t.TempDir(), "does-not-exist.json")

	err := LoadJSON(path, &testDoc{})
	if err == nil {
		t.Fatal("LoadJSON on missing file: want error, got nil")
	}
	if !os.IsNotExist(err) {
		t.Errorf("LoadJSON on missing file: err = %v, want os.IsNotExist(err) == true", err)
	}
}

func TestSave_NoLeftoverTempFiles(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "doc.json")

	if err := Save(path, &testDoc{Name: "Braton", N: 1}); err != nil {
		t.Fatalf("Save: %v", err)
	}

	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("ReadDir: %v", err)
	}
	for _, e := range entries {
		if strings.Contains(e.Name(), ".tmp-") {
			t.Errorf("leftover temp file after Save: %s", e.Name())
		}
	}
}

func TestSave_CreatesOneBackupPerSave(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "doc.json")

	if err := Save(path, &testDoc{Name: "v1", N: 1}); err != nil {
		t.Fatalf("Save: %v", err)
	}

	backups := listBackups(t, dir)
	if len(backups) != 1 {
		t.Fatalf("backups after 1 Save = %v, want exactly 1 file", backups)
	}
}

// TestSave_PrunesBackupsBeyondDefaultBackupKeep seeds the backups directory
// with more than DefaultBackupKeep pre-existing (older) backup files, then
// triggers one more Save, and checks that pruning keeps only the newest
// DefaultBackupKeep files. Backup filenames embed a second-granularity
// timestamp, so this test crafts filenames directly instead of sleeping
// between real Saves (which would make the test slow and timing-flaky).
func TestSave_PrunesBackupsBeyondDefaultBackupKeep(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "doc.json")
	backupsDir := filepath.Join(dir, "backups")
	if err := os.MkdirAll(backupsDir, 0755); err != nil {
		t.Fatalf("MkdirAll backups: %v", err)
	}

	// Seed DefaultBackupKeep older, already-existing backups (older than
	// whatever timestamp the real Save below will produce).
	seeded := []string{
		"doc.20200101-000001.json",
		"doc.20200101-000002.json",
		"doc.20200101-000003.json",
		"doc.20200101-000004.json",
		"doc.20200101-000005.json",
	}
	for _, name := range seeded {
		if err := os.WriteFile(filepath.Join(backupsDir, name), []byte(`{"name":"seed","n":0}`), 0644); err != nil {
			t.Fatalf("seed backup %s: %v", name, err)
		}
	}
	if len(seeded) != DefaultBackupKeep {
		t.Fatalf("test setup bug: seeded %d backups, DefaultBackupKeep = %d", len(seeded), DefaultBackupKeep)
	}

	// This Save adds a 6th (newest) backup, so pruning must remove exactly
	// one — the oldest of the seeded files.
	if err := Save(path, &testDoc{Name: "newest", N: 6}); err != nil {
		t.Fatalf("Save: %v", err)
	}

	backups := listBackups(t, dir)
	if len(backups) != DefaultBackupKeep {
		t.Fatalf("backups after prune = %v (len %d), want exactly %d", backups, len(backups), DefaultBackupKeep)
	}
	for _, b := range backups {
		if b == "doc.20200101-000001.json" {
			t.Errorf("oldest seeded backup should have been pruned, but %s still exists", b)
		}
	}
}

func TestLoadJSON_RecoversFromNewestGoodBackupWhenPrimaryIsCorrupt(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "doc.json")

	// First Save produces a good backup of {Name:"good", N:1}.
	if err := Save(path, &testDoc{Name: "good", N: 1}); err != nil {
		t.Fatalf("Save (good): %v", err)
	}

	// Second Save corrupts the *backup* rotation by first saving a second
	// good version, then we hand-corrupt only the primary file so the most
	// recent backup on disk still holds valid, recoverable data.
	if err := Save(path, &testDoc{Name: "good2", N: 2}); err != nil {
		t.Fatalf("Save (good2): %v", err)
	}

	if err := os.WriteFile(path, []byte(`{"name": "broken", "n": `), 0644); err != nil {
		t.Fatalf("corrupt primary file: %v", err)
	}

	var got testDoc
	if err := LoadJSON(path, &got); err != nil {
		t.Fatalf("LoadJSON: want automatic recovery, got error: %v", err)
	}
	if got.Name != "good2" || got.N != 2 {
		t.Errorf("LoadJSON recovered = %+v, want {Name:good2 N:2} (newest good backup)", got)
	}

	// Self-healing: the primary file should now be rewritten with the
	// recovered content, not left corrupt, so the next launch doesn't need
	// to recover again.
	healedData, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read healed primary file: %v", err)
	}
	var healed testDoc
	if err := json.Unmarshal(healedData, &healed); err != nil {
		t.Fatalf("healed primary file is not valid JSON: %v (content: %s)", err, healedData)
	}
	if healed.Name != "good2" || healed.N != 2 {
		t.Errorf("healed primary file = %+v, want {Name:good2 N:2}", healed)
	}
}

func TestLoadJSON_ReturnsParseErrorWhenNoBackupIsUsable(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "doc.json")

	// No prior Save, so there is no backups/ directory at all — only a
	// hand-written corrupt primary file.
	if err := os.WriteFile(path, []byte(`not json`), 0644); err != nil {
		t.Fatalf("write corrupt primary file: %v", err)
	}

	var got testDoc
	err := LoadJSON(path, &got)
	if err == nil {
		t.Fatal("LoadJSON with corrupt primary and no backups: want error, got nil")
	}
	if !strings.Contains(err.Error(), "no usable backup found") {
		t.Errorf("LoadJSON error = %q, want it to mention \"no usable backup found\"", err.Error())
	}
}

// listBackups returns the sorted basenames of files under dir/backups.
func listBackups(t *testing.T, dir string) []string {
	t.Helper()
	entries, err := os.ReadDir(filepath.Join(dir, "backups"))
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		t.Fatalf("ReadDir backups: %v", err)
	}
	var names []string
	for _, e := range entries {
		names = append(names, e.Name())
	}
	sort.Strings(names)
	return names
}

File: ts/server/persist.test.ts
// Port of pkg/persist/persist_test.go — same 7 cases, same names (translated
// to TS test-naming convention), same assertions. This is the executable
// spec for the persist layer; written test-first per the migration plan.

import { afterEach, beforeEach, expect, test } from "bun:test";
import { z } from "zod";
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import { DEFAULT_BACKUP_KEEP, NotFoundError, loadJSON, saveJSON } from "./persist.ts";

const testDocSchema = z.object({ name: z.string(), n: z.number() });
type TestDoc = z.infer<typeof testDocSchema>;

let dir: string;
beforeEach(async () => {
  dir = await fs.mkdtemp(path.join(os.tmpdir(), "wsg-persist-test-"));
});
afterEach(async () => {
  await fs.rm(dir, { recursive: true, force: true });
});

async function listBackups(d: string): Promise<string[]> {
  try {
    const entries = await fs.readdir(path.join(d, "backups"));
    return entries.sort();
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return [];
    throw err;
  }
}

test("saveJSON + loadJSON round trip", async () => {
  const docPath = path.join(dir, "doc.json");

  await saveJSON(docPath, { name: "Ash Prime", n: 3 } satisfies TestDoc);

  const got = await loadJSON(docPath, testDocSchema);
  expect(got.name).toBe("Ash Prime");
  expect(got.n).toBe(3);
});

test("loadJSON on missing file preserves not-found-ness", async () => {
  const docPath = path.join(dir, "does-not-exist.json");

  await expect(loadJSON(docPath, testDocSchema)).rejects.toThrow();
  try {
    await loadJSON(docPath, testDocSchema);
    throw new Error("expected loadJSON to throw");
  } catch (err) {
    expect(err).toBeInstanceOf(NotFoundError);
    expect((err as NotFoundError).code).toBe("ENOENT");
  }
});

test("saveJSON leaves no leftover temp files", async () => {
  const docPath = path.join(dir, "doc.json");

  await saveJSON(docPath, { name: "Braton", n: 1 } satisfies TestDoc);

  const entries = await fs.readdir(dir);
  for (const name of entries) {
    expect(name.includes(".tmp-")).toBe(false);
  }
});

test("saveJSON creates exactly one backup per save", async () => {
  const docPath = path.join(dir, "doc.json");

  await saveJSON(docPath, { name: "v1", n: 1 } satisfies TestDoc);

  const backups = await listBackups(dir);
  expect(backups.length).toBe(1);
});

// Seeds the backups directory with more than DEFAULT_BACKUP_KEEP pre-existing
// (older) backup files, then triggers one more saveJSON, and checks that
// pruning keeps only the newest DEFAULT_BACKUP_KEEP files. Backup filenames
// embed a second-granularity timestamp, so this crafts filenames directly
// instead of sleeping between real saves (which would make the test slow and
// timing-flaky) — same approach as the Go original.
test("saveJSON prunes backups beyond DEFAULT_BACKUP_KEEP", async () => {
  const docPath = path.join(dir, "doc.json");
  const backupsDir = path.join(dir, "backups");
  await fs.mkdir(backupsDir, { recursive: true });

  const seeded = [
    "doc.20200101-000001.json",
    "doc.20200101-000002.json",
    "doc.20200101-000003.json",
    "doc.20200101-000004.json",
    "doc.20200101-000005.json",
  ];
  expect(seeded.length).toBe(DEFAULT_BACKUP_KEEP);
  for (const name of seeded) {
    await fs.writeFile(path.join(backupsDir, name), `{"name":"seed","n":0}`, "utf8");
  }

  // This save adds a 6th (newest) backup, so pruning must remove exactly
  // one — the oldest of the seeded files.
  await saveJSON(docPath, { name: "newest", n: 6 } satisfies TestDoc);

  const backups = await listBackups(dir);
  expect(backups.length).toBe(DEFAULT_BACKUP_KEEP);
  expect(backups.includes("doc.20200101-000001.json")).toBe(false);
});

test("loadJSON recovers from newest good backup when primary is corrupt", async () => {
  const docPath = path.join(dir, "doc.json");

  // First save produces a good backup of {name:"good", n:1}.
  await saveJSON(docPath, { name: "good", n: 1 } satisfies TestDoc);
  // Second save produces a newer good backup of {name:"good2", n:2}, then we
  // hand-corrupt only the primary file so the most recent backup on disk
  // still holds valid, recoverable data.
  await saveJSON(docPath, { name: "good2", n: 2 } satisfies TestDoc);
  await fs.writeFile(docPath, `{"name": "broken", "n": `, "utf8");

  const got = await loadJSON(docPath, testDocSchema);
  expect(got.name).toBe("good2");
  expect(got.n).toBe(2);

  // Self-healing: the primary file should now be rewritten with the
  // recovered content, not left corrupt, so the next launch doesn't need to
  // recover again.
  const healedRaw = await fs.readFile(docPath, "utf8");
  const healed = testDocSchema.parse(JSON.parse(healedRaw));
  expect(healed.name).toBe("good2");
  expect(healed.n).toBe(2);
});

test("loadJSON returns parse error when no backup is usable", async () => {
  const docPath = path.join(dir, "doc.json");

  // No prior save, so there is no backups/ directory at all — only a
  // hand-written corrupt primary file.
  await fs.writeFile(docPath, "not json", "utf8");

  await expect(loadJSON(docPath, testDocSchema)).rejects.toThrow(/no usable backup found/);
});
